Looppoint

studyztp opened the pull request:
This implementation allow gem5 to take in and keep track on the input list of target PCs and their target counts. It can raise an exit event when the PC counts of the target PC reaches the target count during the simulation. With the exit events, we can take and restore checkpoints to do detail analysis on the selected regions. 
The implementation of this is mainly based on the Probe object in gem5. The Probe object is designed for custom tracking. In the BaseCPU, there is a ProbePoint that keeps track of committed instruction PC. This ProbePoint will notify its ProbeListener every time a core finishes committing. The LoopPoint is a ProbeListnerObject that’s receiving and handling the notification from each core. Because of the design of the Probe, one ProbeListener can only listen to one SimObject’s ProbePoints. Therefore, there is a LoopPointManager SimObject to handle the PC counting for multithread.

studyztp committed:
stdlib: sth sth sth

Change-Id: I03d92ec63068d19547197583cf713136ae52d627

studyztp committed:
util: added looppoint and looppointmanager

added looppoint probelistenerobject to track probepoint from basecpu
added looppointmanager to enable multithread pc counting

Change-Id: Ib9d2dee442a719b8947ab6a2938f44c9e6d07a31

studyztp committed:
util: added printf to verify the counts when doing multithreads

studyztp committed:
util: allow input list of targetPC and targetCount

studyztp committed:
util: allow detecting different targetCount for same PC

Now LoopPointManager stores targetCount and PCcount as a hash map
It has an outstream to output information about each exit event
LoopPoint now calls the LoopPointManager check_pc() with pc instead
of index

powerjg commented on src/cpu/BaseCPU.py:
```diff
@@ -315,3 +317,10 @@ def __init__(self, **kwargs):
                 + ["interrupts[0].int_requestor"]
             )
+
+    def addLoopPointProbe(self, targetpc, manager):
```
Let's put this in the stdlib instead of in the SimObject description file. I'd like to start migrating things out of this file into the stdlib more generally :)

powerjg commented on src/cpu/probes/LoopPoint.py:
```diff
@@ -0,0 +1,13 @@
+from m5.params import *
```
Needs copyright

powerjg commented on src/cpu/probes/LoopPoint.py:
```diff
@@ -0,0 +1,13 @@
+
+    target_pc = VectorParam.UInt64(0, "the target PC")
+    core = Param.BaseCPU(NULL, "the connected cpu")
+    LPmanager = Param.LoopPointManager(NULL, "the looppoint manager")
```
```suggestion
    manager = Param.LoopPointManager(NULL, "the looppoint manager")
```

I'd just call this the `manager`. This way you don't have to violate the style with a capital first letter.

Also, what does it mean to be NULL here? Is that valid? If not, then we should just have no default.

studyztp replied:
I got python error when building gem5 if I don't put NULL here. I think NULL is valid when Param is a SimObject.

powerjg replied:
Odd! I would not expect that error. @BobbyRBruce, we should document this behavior or we should figure out if it's a bug. It surprises me that a default value is *required* for SimObject params 

What I meant by valid is whether or not it would be valid for the user to specify NULL. I think not. I believe there would be a fatal or segfault raised. That said, if gem5 requires a default, then that's how it is.

studyztp replied:
It could be I was doing something wrong with other parts. I'll test with only the Param SimObject without default value now.

powerjg commented on src/cpu/probes/LoopPoint.py:
```diff
@@ -0,0 +1,13 @@
+    cxx_class = "gem5::LoopPoint"
+
+    target_pc = VectorParam.UInt64(0, "the target PC")
+    core = Param.BaseCPU(NULL, "the connected cpu")
```
Also, what does it mean to be NULL here? Is that valid? If not, then we should just have no default.

powerjg commented on src/cpu/probes/LoopPoint.py:
```diff
@@ -0,0 +1,13 @@
+    cxx_header = "cpu/probes/looppoint.hh"
+    cxx_class = "gem5::LoopPoint"
+
+    target_pc = VectorParam.UInt64(0, "the target PC")
```
What does it mean to be 0 here? Is that valid? If not, then we should just have no default.

studyztp replied:
I changed the default value from ```0``` to ```[]```. ```[]``` should be the valid default value for VectorParam. If I don't set a default value, it raised python errors when I tried to build gem5.

powerjg replied:
Like the comment above, this should not be the case. Maybe it is, but I think it's a bug in gem5.

I can find plenty of examples both where there is no default for the `Param.<SimObject>` and when there's no default for a `VectorParam`. 

I'm not saying I don't believe you, but it's very strange! :)

powerjg commented on src/cpu/probes/LoopPointManager.py:
```diff
@@ -0,0 +1,16 @@
+    cxx_header = "cpu/probes/looppointmanager.hh"
+    cxx_class = "gem5::LoopPointManager"
+
+    target_count = VectorParam.Int(0, "the target PC count")
+    target_pc = VectorParam.UInt64(0, "the target PC")
```
Probably shouldn't have any default values

powerjg commented on src/cpu/probes/LoopPointManager.py:
```diff
@@ -0,0 +1,16 @@
+from m5.params import *
```
Copyright

powerjg commented on src/cpu/probes/SConscript:
```diff
@@ -0,0 +1,8 @@
+Import('*')
```
copyright

powerjg commented on src/cpu/probes/looppoint.cc:
```diff
@@ -0,0 +1,42 @@
+#include "cpu/probes/looppoint.hh"
```
copyright

powerjg commented on src/cpu/probes/looppoint.cc:
```diff
@@ -0,0 +1,42 @@
+    targetPC(p.target_pc),
+    cpuptr(p.core),
+    lpmanager(p.LPmanager)
+{
```
If it's required that certain things are non-null, this is a good place to throw a `fatal` if they are null.

powerjg commented on src/cpu/probes/looppoint.cc:
```diff
@@ -0,0 +1,42 @@
+void
+LoopPoint::check_pc(const Addr& pc)
+{
+    for (int i =0; i<targetPC.size();i++)
+    {
+        if(pc == targetPC[i])
+        {
+            if (lpmanager->check_count(pc))
+                cpuptr->scheduleInstStop(0,1,"simpoint starting point found");
+        }
+    }
```
```suggestion
    for (int i =0; i < targetPC.size(); i++) {
        if (pc == targetPC[i]) {
            if (lpmanager->check_count(pc)) {
                cpuptr->scheduleInstStop(0,1,"simpoint starting point found");
            }
        }
    }
```

powerjg commented on src/cpu/probes/looppoint.hh:
```diff
@@ -0,0 +1,30 @@
+#ifndef __CPU_PROBES_LOOPPOINT_HH__
```
copyright

powerjg commented on src/cpu/probes/looppoint.hh:
```diff
@@ -0,0 +1,30 @@
+        void check_pc(const Addr&);
+
+    private:
+        const std::vector<Addr> targetPC;
```
I suggest making this an `unordered_map<Addr,unsigned>`. It should perform better and you won't have to use `check_pc`. Instead, you can set this up at initialization time (maybe through `init()`?)

powerjg replied:
Actually, I think this should be an `unordered_set`

powerjg commented on src/cpu/probes/looppointmanager.cc:
```diff
@@ -0,0 +1,57 @@
+#include "cpu/probes/looppointmanager.hh"
```
copyright

powerjg commented on src/cpu/probes/looppointmanager.cc:
```diff
@@ -0,0 +1,57 @@
+LoopPointManager::LoopPointManager(const LoopPointManagerParams &p)
+    : SimObject(p)
+{
+    for (int i = 0; i< p.target_pc.size(); i++)
+    {
+        auto map_itr = targetCount.find(p.target_pc[i]);
+        if (map_itr == targetCount.end()){
+            std::vector<int> pcCount = {p.target_count[i]};
+            targetCount.insert(std::make_pair(p.target_pc[i], pcCount));
+            counter.insert(std::make_pair(p.target_pc[i],0));
+        } else {
+            std::vector<int>& pcCount = map_itr->second;
+            pcCount.push_back(p.target_count[i]);
+        }
+    }
+    info = simout.create("LoopPointInfo.txt", false);
+    if (!info)
+        fatal("unable to open LoopPoint info txt");
```
```suggestion
    for (int i = 0; i< p.target_pc.size(); i++) {
        auto map_itr = targetCount.find(p.target_pc[i]);
        if (map_itr == targetCount.end()) {
            std::vector<int> pcCount = {p.target_count[i]};
            targetCount.insert(std::make_pair(p.target_pc[i], pcCount));
            counter.insert(std::make_pair(p.target_pc[i],0));
        } else {
            std::vector<int>& pcCount = map_itr->second;
            pcCount.push_back(p.target_count[i]);
        }
    }
    info = simout.create("LoopPointInfo.txt", false);
    if (!info) {
        fatal("unable to open LoopPoint info txt");
    }
```

powerjg commented on src/cpu/probes/looppoint.hh:
```diff
@@ -0,0 +1,30 @@
+
+class LoopPoint : public ProbeListenerObject
+{
+    public:
```
Two spaces instead of 4.

powerjg commented on src/cpu/probes/looppointmanager.cc:
```diff
@@ -0,0 +1,57 @@
+LoopPointManager::LoopPointManager(const LoopPointManagerParams &p)
+    : SimObject(p)
+{
+    for (int i = 0; i< p.target_pc.size(); i++)
```
Can you add some comments here explaining this code?

powerjg commented on src/cpu/probes/looppointmanager.cc:
```diff
@@ -0,0 +1,57 @@
+    count += 1;
+    auto target_itr = targetCount.find(pc);
+    std::vector<int>& targetcount = target_itr->second;
+    for (std::vector<int>::iterator iter = targetcount.begin(); iter < targetcount.end(); iter++)
+    {
+        if(*iter==count)
+        {
+            targetcount.erase(iter);
+            *info->stream() << curTick() << " : " << pc << " : " << count << " \n "; 
+            return true;
+        }
+    }
```
```suggestion
    for (std::vector<int>::iterator iter = targetcount.begin(); iter < targetcount.end(); iter++) {
        if(*iter==count) {
            targetcount.erase(iter);
            *info->stream() << curTick() << " : " << pc << " : " << count << " \n "; 
            return true;
        }
    }
```

powerjg commented on src/cpu/probes/looppointmanager.cc:
```diff
@@ -0,0 +1,57 @@
+        if(*iter==count)
+        {
+            targetcount.erase(iter);
+            *info->stream() << curTick() << " : " << pc << " : " << count << " \n "; 
```
I would exit here. I think you can use a function like `exitSimLoop`. See `exit_event.hh` or something like that.

powerjg commented on src/cpu/probes/looppointmanager.hh:
```diff
@@ -0,0 +1,32 @@
+#ifndef __CPU_PROBES_LOOPPOINTMANAGER_HH__
```
copyright

powerjg commented on src/cpu/probes/looppointmanager.hh:
```diff
@@ -0,0 +1,32 @@
+    private:
+        // const std::vector<int> targetCount;
+        // std::vector<int> counter;
+        OutputStream *info;
+        std::unordered_map<uint64_t, std::vector<int>> targetCount;
+        std::unordered_map<uint64_t,int> counter;
```
Add some comments explaining these.

powerjg commented on src/cpu/probes/looppointmanager.hh:
```diff
@@ -0,0 +1,32 @@
+        LoopPointManager(const LoopPointManagerParams &params);
+        virtual ~LoopPointManager();
+        virtual void init();
+        bool check_count(uint64_t);
```
Docstring and also give the variable a name :)

studyztp committed:
stdlib: added LoopPointProbe connection function

Moved AddLoopPointProbe() from BaseCPU to abstract_core and base_cpu_core
Changed LPmanager to manager in looppoint.hh

studyztp committed:
cpu: added copyright to all LoopPoint related files

studyztp committed:
util: Fixed default value and Changed targetPC type

Changed targetPC in looppoint from vector to unordered_set
Fixed the default value for py files

studyztp committed:
util: Use exitSimLoopNow to exit

Let LoopPointManager to raise exitevent instead of LoopPoint with
exitSimLoopNow().
Changed the uint64_t type to gem5::Addr to match the style

studyztp committed:
util: Fixed the style issues and added fatal for NULL input

studyztp committed:
util: added comments

studyztp committed:
util: added pydocs and partially added param type

studyztp commented:
@powerjg Thank you for the feedbacks! I fixed the issues and added the comments.

studyztp committed:
util: removed all defualt values in LoopPoint.py and LoopPointManager.py

studyztp committed:
stdlib: added LoopPoint stdlib module to process input file

studyztp committed:
stdlib, probe: added relative count for warmup+detail checkpoint

studyztp committed:
util: don't target PC == 0

studyztp committed:
util: output index

studyztp committed:
util: removed index in output but added exit when targetCount is empty

studyztp committed:
util,stdlib: relative_count in LoopPointManager;take in output file;

In LoopPointManager, it now takes in a relative_count list to calculate
and output the relative data we need for restoring to the LoopPointInfo
output file.

In LoopPoints(stdlib) module, the setup_restore() function is added
to use the output file described above to setup PC track for restoring
checkpoints taken for LoopPoint simulation region (warmup+detail)

studyztp committed:
util,stdlib: fixed conflict exit events after all targets encountered

In LoopPointManager, the original appoarch raised conflict exit events
when all targets were encountered. Now it raises SIMPOINT_BEGIN with
exitSimLoopNow() and raises EXIT with exitSimLoop to fix this issue.

In stdlib LoopPoints, it now has a temporary testing function.
It doesn't take in anything when creating a LoopPoints module object,
instead, it takes in files depend on taking checkpoints or restoring
checkpoints.

studyztp committed:
stdlib: added the 'break' for a loop

studyztp committed:
stdlib,util: able to restore warmup+detail checkpoint with output file

In LoopPointManger, added default value for related_count
In LoopPoints, fixed some issues and added some prints for debugging

studyztp committed:
stdlib, probe: added region id to LoopPointInfo.txt

The region id are correctly assigned in the corresponding checkpoint
line.

studyztp committed:
stdlib, probe: changed stdlib LoopPoint hierachy

In src/python/gem5/util/looppoint.py
Separated the LoopPoint stdlib class into LoopPointsRestore and
LoopPointsCheckPoint. They both inhereited from an abstract class called
BaseLoopPoints.

In the LoopPointManager.py
removed default values for all params

In base_cpu_core.py
added type to its params

studyztp committed:
stdlib: added type for BaseLoopPoints init function

studyztp committed:
stdlib: added looppoint to se_binary_workload

powerjg commented on src/cpu/probes/LoopPoint.py:
```diff
@@ -0,0 +1,39 @@
+    cxx_header = "cpu/probes/looppoint.hh"
+    cxx_class = "gem5::LoopPoint"
+
+    target_pc = VectorParam.UInt64("the target PC")
```
```suggestion
    target_pc = VectorParam.Addr("the target PC")
```

Is `Addr` a valid param type? I think it is.

powerjg commented on src/cpu/probes/LoopPointManager.py:
```diff
@@ -0,0 +1,49 @@
+    cxx_class = "gem5::LoopPointManager"
+
+    target_count = VectorParam.Int("the target PC count")
+    target_pc = VectorParam.UInt64("the target PC")
```
```suggestion
    target_pc = VectorParam.Addr("the target PC")
```

powerjg commented on src/cpu/probes/LoopPointManager.py:
```diff
@@ -0,0 +1,49 @@
+    target_count = VectorParam.Int("the target PC count")
+    target_pc = VectorParam.UInt64("the target PC")
+    region_id = VectorParam.Int("the simulation region id")
+    relative_pc = VectorParam.UInt64("the relative PC of the target PC")
```
```suggestion
    relative_pc = VectorParam.Addr("the relative PC of the target PC")
```

powerjg commented on src/cpu/probes/LoopPointManager.py:
```diff
@@ -0,0 +1,49 @@
+from m5.params import *
+from m5.objects import SimObject
+
+class LoopPointManager(SimObject):
```
Small nit: I would put this class in the `LoopPoint.py` file.

powerjg commented on src/python/gem5/utils/looppoint.py:
```diff
@@ -0,0 +1,219 @@
+        
+class BaseLoopPoints:
+    
+    __metaclass__ = ABCMeta
```
I don't think this is needed.

powerjg commented on src/python/gem5/components/processors/base_cpu_core.py:
```diff
@@ -165,3 +170,15 @@ def set_inst_stop_any_thread(self, inst: int, init: bool) -> None:
+    @overrides(AbstractCore)
+    def addLoopPointProbe(
+        self, 
+        targetpc: List[AbstractCore],
```
```suggestion
        targetpc: List[int],
```

powerjg commented on src/python/gem5/components/processors/base_cpu_core.py:
```diff
@@ -25,6 +25,8 @@
 
 from typing import Optional, List
+
+from m5.params import UInt64
```
Remove

powerjg commented on src/cpu/probes/LoopPointManager.py:
```diff
@@ -0,0 +1,49 @@
+    relative_pc = VectorParam.UInt64("the relative PC of the target PC")
+    relative_count = VectorParam.Int("the relative PC count of the target PC")
+
+    def setup(self, cpulist):
```
Is this function required? I would rather not have general python in the "SimObject declaration files" if it's avoidable.

powerjg commented on src/python/gem5/utils/looppoint.py:
```diff
@@ -0,0 +1,219 @@
+        self._looppointManager.region_id = self._regionID
+        self._looppointManager.relative_pc = self._relativePC
+        self._looppointManager.relative_count = self._relativeCount
+        self._looppointManager.setup(cpuList)
```
```suggestion
    def setup_processor(
        self, 
        processor: List[AbstractProcessor], 
        ) -> None:
        self._looppointManager.target_count = self._checkpointCount
        self._looppointManager.target_pc = self._checkpointPC
        self._looppointManager.region_id = self._regionID
        self._looppointManager.relative_pc = self._relativePC
        self._looppointManager.relative_count = self._relativeCount
        for core in processor.get_cores():
            core.addLoopPointProbe(self._checkpointPC + self._relativePC, self._looppointManager)
```

Note, I'm not very confident on the last line.

Whether you go with this suggestion or something closer to the original... I suggest using the name "core" or "processor" and avoiding the name "cpu" for variables.

studyztp committed:
stdlib, probe: added comments and fixed some styling

studyztp committed:
sim: Added LoopPoint probes and multicore manager

LoopPoint is a multithread sampling method that targets PC and PC count

The LoopPoint probe object is used to keep track of the targeting PCs
and notice the LoopPointManager when the current PC is the targeting
PC.
The LoopPointManager is a SimObject that stores all the LoopPoints
related information and raise exit event when the targeting PC gets to
the targeting count.

Change-Id: Id2ab1b606e4a13a21fb79664243d4f140392d7b4

powerjg commented on src/cpu/probes/looppoint.cc:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+LoopPoint::~LoopPoint()
+{}
+
+void
+LoopPoint::init()
+{}
```
You can remove these empty functions

powerjg commented on src/cpu/probes/looppoint.hh:
```diff
@@ -0,0 +1,76 @@
+{
+  public:
+    LoopPoint(const LoopPointParams &params);
+    virtual ~LoopPoint();
+    virtual void init();
```
Not needed :)

powerjg commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -111,3 +112,37 @@ def set_se_binary_workload(
         # Set whether to exit on work items for the se_workload
         self.exit_on_work_items = exit_on_work_items
+
+    def set_se_looppoint_workload(
```
I think we changed the way we're doing simpoints. Can you follow what's done here; https://gem5-review.googlesource.com/c/public/gem5/+/64432

powerjg reviewed (commented):
A few comments. I think you might have closed this, but the comments still stand :).

Good work so far!

studyztp commented:
I'll fix them and move them to Gerrit. Thanks for the great comments! :)

studyztp committed:
stdlib, sim: merged LoopPointManager.py& removed setup

Merged LoopPointManager.py into LoopPoint.py and removed the setup
function.
Changed the UInt64 to Addr Param type in the SimObject file.
Now BaseLoopPoint is responsible for connecting all the LoopPoint
information with all the running cores.

studyztp committed:
stdlib: changed setup_cpu to setup_process

The original setup_cpu function in BaseLoopPoint now takes in processor
instead of a list of cores.

studyztp committed:
stdlib: Added stdlib LoopPoint classes

The stdlib LoopPoint classes have to work with the LoopPoint probes and
the LoopPointManger. They are all added for running the LoopPoint
sampling method.
The stdlib LoopPoint has three classes, an abstract class and two
subclasses.
The LoopPointRestore is responsible for profiling restoring related
points.
The LoopPointCheckpoint is responsible for profiling checkpoints.
The BaseLoopPoint stores the general LoopPoints information and setup
the LoopPointManager and LoopPoint probes to each core.

Change-Id: Ie10d11e6b33189245997d3ab8e9f75cf03d76a39

studyztp committed:
stdlib: Allow se_binary_workload to setup LoopPoints

Added a set_se_looppoint_workload function to take in information for
workload and a stdlib LoopPoint object that stores all the information
the workload needed to run the LoopPoint sampling method.
Added a get_looppoint function to return the stdlib LoopPoint object.

Change-Id: I7872d594bf4c86bc5d9329319ec0be72b2b9670d

studyztp committed:
ran the precommit program on them

studyztp committed:
Merge with the gerrit version

studyztp committed:
Merge with the gerrit version

studyztp committed:
Merge branch 'looppoint' of https://github.com/darchr/gem5 into looppoint
